File: native/lexer/lexer.go
// Package lexer is the first native Koschei compiler slice.
//
// It is dependency-free, deterministic, UTF-8 aware, and fail-closed on source
// and token budgets. It does not execute code, import packages, or touch ambient
// authority. The current Python frontend remains the compatibility oracle until
// native parity is complete.
package lexer

import (
	"errors"
	"fmt"
	"unicode"
	"unicode/utf8"
)

const (
	DefaultMaxSourceBytes = 4 << 20
	DefaultMaxTokens      = 1_000_000
)

var (
	ErrInvalidUTF8    = errors.New("source is not valid UTF-8")
	ErrSourceTooLarge = errors.New("source exceeds configured byte budget")
	ErrTokenBudget    = errors.New("token budget exhausted")
)

type Config struct {
	KeepComments   bool
	MaxSourceBytes int
	MaxTokens      int
}

func (config Config) normalized() Config {
	if config.MaxSourceBytes == 0 {
		config.MaxSourceBytes = DefaultMaxSourceBytes
	}
	if config.MaxTokens == 0 {
		config.MaxTokens = DefaultMaxTokens
	}
	return config
}

type Error struct {
	Line    int
	Column  int
	Message string
	Cause   error
}

func (failure *Error) Error() string {
	return fmt.Sprintf("[line %d, column %d] %s", failure.Line, failure.Column, failure.Message)
}

func (failure *Error) Unwrap() error { return failure.Cause }

type Lexer struct {
	source      []rune
	config      Config
	start       int
	current     int
	line        int
	column      int
	startLine   int
	startColumn int
	tokens      []Token
}

var keywords = map[string]Kind{
	"fn": FN, "let": LET, "mut": MUT, "or": OR, "return": RETURN,
	"if": IF, "else": ELSE, "while": WHILE, "for": FOR, "in": IN,
	"struct": STRUCT, "enum": ENUM, "match": MATCH, "import": IMPORT,
	"true": TRUE, "false": FALSE,
}

var builtinTypes = map[string]struct{}{
	"SystemCaps": {}, "NetRoot": {}, "DiskRoot": {}, "EnvRoot": {},
	"ProcessRoot": {}, "NetCaps": {}, "DiskCaps": {}, "DiskReadCaps": {},
	"EnvCaps": {}, "ProcessCaps": {}, "List": {}, "Map": {}, "String": {},
	"Int": {}, "Float": {}, "Bool": {}, "Void": {}, "Error": {},
	"Result": {}, "Option": {},
}

var single = map[rune]Kind{
	'(': LEFTPAREN, ')': RIGHTPAREN, '{': LEFTBRACE, '}': RIGHTBRACE,
	',': COMMA, ':': COLON, '.': DOT, '+': PLUS, '*': STAR,
	';': SEMICOLON, '[': LEFTBRACKET, ']': RIGHTBRACKET,
}

var escapes = map[rune]rune{
	'n': '\n', 'r': '\r', 't': '\t', '"': '"', '\\': '\\', '{': '{', '}': '}',
}

func Tokenize(source string, config Config) ([]Token, error) {
	config = config.normalized()
	if config.MaxSourceBytes < 1 {
		return nil, &Error{Line: 1, Column: 1, Message: "max source byte budget must be positive", Cause: ErrSourceTooLarge}
	}
	if config.MaxTokens < 1 {
		return nil, &Error{Line: 1, Column: 1, Message: "max token budget must be positive", Cause: ErrTokenBudget}
	}
	if len(source) > config.MaxSourceBytes {
		return nil, &Error{Line: 1, Column: 1, Message: fmt.Sprintf("source exceeds %d-byte budget", config.MaxSourceBytes), Cause: ErrSourceTooLarge}
	}
	if !utf8.ValidString(source) {
		return nil, &Error{Line: 1, Column: 1, Message: ErrInvalidUTF8.Error(), Cause: ErrInvalidUTF8}
	}
	machine := &Lexer{
		source: []rune(source), config: config, line: 1, column: 1,
		tokens: make([]Token, 0, min(len(source)/3+1, config.MaxTokens)),
	}
	return machine.tokenize()
}

func (lexer *Lexer) tokenize() ([]Token, error) {
	for !lexer.atEnd() {
		lexer.start = lexer.current
		lexer.startLine = lexer.line
		lexer.startColumn = lexer.column
		if err := lexer.scan(); err != nil {
			return nil, err
		}
	}
	if err := lexer.add(Token{Kind: EOF, Line: lexer.line, Column: lexer.column}); err != nil {
		return nil, err
	}
	return lexer.tokens, nil
}

func (lexer *Lexer) scan() error {
	char := lexer.advance()
	if char == ' ' || char == '\r' || char == '\t' || char == '\n' {
		return nil
	}
	if char == '/' {
		if lexer.match('/') {
			for lexer.peek() != '\n' && lexer.peek() != 0 {
				lexer.advance()
			}
			if lexer.config.KeepComments {
				text := lexer.slice()
				return lexer.emit(COMMENT, text, text, nil)
			}
			return nil
		}
		return lexer.emit(SLASH, "/", "/", nil)
	}
	if char == '"' {
		return lexer.stringValue()
	}
	if unicode.IsDigit(char) {
		return lexer.number()
	}
	if unicode.IsLetter(char) || char == '_' {
		return lexer.identifier()
	}
	switch char {
	case '-':
		if lexer.match('>') {
			return lexer.emit(ARROW, "->", "->", nil)
		}
		return lexer.emit(MINUS, "-", "-", nil)
	case '=':
		if lexer.match('>') {
			return lexer.emit(FATARROW, "=>", "=>", nil)
		}
		if lexer.match('=') {
			return lexer.emit(EQUALEQUAL, "==", "==", nil)
		}
		return lexer.emit(EQUAL, "=", "=", nil)
	case '!':
		if lexer.match('=') {
			return lexer.emit(BANGEQUAL, "!=", "!=", nil)
		}
		return lexer.emit(BANG, "!", "!", nil)
	case '&':
		if lexer.match('&') {
			return lexer.emit(AMPAMP, "&&", "&&", nil)
		}
		return lexer.failure("unexpected '&'; use '&&' for logical and", nil)
	case '|':
		if lexer.match('|') {
			return lexer.emit(PIPEPIPE, "||", "||", nil)
		}
		return lexer.failure("unexpected '|'; use '||' for logical or", nil)
	case '<':
		if lexer.match('=') {
			return lexer.emit(LESSEQUAL, "<=", "<=", nil)
		}
		return lexer.emit(LESS, "<", "<", nil)
	case '>':
		if lexer.match('=') {
			return lexer.emit(GREATEREQUAL, ">=", ">=", nil)
		}
		return lexer.emit(GREATER, ">", ">", nil)
	}
	if kind, ok := single[char]; ok {
		text := string(char)
		return lexer.emit(kind, text, text, nil)
	}
	return lexer.failure(fmt.Sprintf("invalid character: %q", char), nil)
}

// isPythonAlnum mirrors Python str.isalnum for identifier continuations:
// Unicode letters plus every Unicode numeric category (Nd, Nl, and No).
func isPythonAlnum(char rune) bool {
	return unicode.IsLetter(char) || unicode.IsNumber(char)
}

func (lexer *Lexer) identifier() error {
	for isPythonAlnum(lexer.peek()) || lexer.peek() == '_' {
		lexer.advance()
	}
	text := lexer.slice()
	kind, ok := keywords[text]
	if !ok {
		_, builtin := builtinTypes[text]
		first, _ := utf8.DecodeRuneInString(text)
		if builtin || unicode.IsUpper(first) {
			kind = TYPE
		} else {
			kind = IDENTIFIER
		}
	}
	return lexer.emit(kind, text, text, nil)
}

func (lexer *Lexer) number() error {
	for unicode.IsDigit(lexer.peek()) {
		lexer.advance()
	}
	if lexer.peek() == '.' && unicode.IsDigit(lexer.peekNext()) {
		lexer.advance()
		for unicode.IsDigit(lexer.peek()) {
			lexer.advance()
		}
	}
	text := lexer.slice()
	return lexer.emit(NUMBER, text, text, nil)
}

func (lexer *Lexer) stringValue() error {
	segments := make([]Segment, 0, 4)
	text := make([]rune, 0, 32)
	flush := func() {
		if len(text) != 0 {
			segments = append(segments, Segment{Kind: "text", Value: string(text)})
			text = text[:0]
		}
	}
	for !lexer.atEnd() {
		char := lexer.advance()
		switch char {
		case '"':
			flush()
			for _, segment := range segments {
				if segment.Kind == "expr" {
					return lexer.emit(STRINGINTERP, lexer.slice(), "", segments)
				}
			}
			value := ""
			if len(segments) != 0 {
				value = segments[0].Value
			}
			return lexer.emit(STRING, lexer.slice(), value, nil)
		case '\\':
			if lexer.atEnd() {
				return lexer.failure("unfinished escape sequence", nil)
			}
			escaped := lexer.advance()
			decoded, ok := escapes[escaped]
			if !ok {
				return lexer.failure(fmt.Sprintf("invalid escape sequence: \\%c", escaped), nil)
			}
			text = append(text, decoded)
		case '{':
			expression, err := lexer.interpolationExpression()
			if err != nil {
				return err
			}
			if expression == "" {
				return lexer.failure("empty interpolation '{}' is invalid", nil)
			}
			flush()
			segments = append(segments, Segment{Kind: "expr", Value: expression})
		case '}':
			return lexer.failure("single '}' in a string is invalid; use '\\}'", nil)
		default:
			text = append(text, char)
		}
	}
	return lexer.failure("unterminated string value", nil)
}

func (lexer *Lexer) interpolationExpression() (string, error) {
	depth := 1
	expression := make([]rune, 0, 32)
	inString := false
	escaped := false
	for !lexer.atEnd() {
		char := lexer.advance()
		if inString {
			expression = append(expression, char)
			if escaped {
				escaped = false
			} else if char == '\\' {
				escaped = true
			} else if char == '"' {
				inString = false
			} else if char == '\n' {
				return "", lexer.failure("string inside interpolation cannot cross a line", nil)
			}
			continue
		}
		switch char {
		case '"':
			inString = true
			expression = append(expression, char)
		case '{':
			depth++
			expression = append(expression, char)
		case '}':
			depth--
			if depth == 0 {
				return trimSpace(string(expression)), nil
			}
			expression = append(expression, char)
		case '\n':
			return "", lexer.failure("interpolation must close with '}' on the same line", nil)
		default:
			expression = append(expression, char)
		}
	}
	return "", lexer.failure("interpolation must close with '}'", nil)
}

// isPythonSpace mirrors the whitespace set used by Python str.strip. Go's
// unicode.IsSpace matches it except for the four information separators below.
func isPythonSpace(char rune) bool {
	return unicode.IsSpace(char) || (char >= '\u001c' && char <= '\u001f')
}

func trimSpace(value string) string {
	runes := []rune(value)
	start, end := 0, len(runes)
	for start < end && isPythonSpace(runes[start]) {
		start++
	}
	for end > start && isPythonSpace(runes[end-1]) {
		end--
	}
	return string(runes[start:end])
}

func (lexer *Lexer) emit(kind Kind, lexeme, value string, segments []Segment) error {
	return lexer.add(Token{Kind: kind, Lexeme: lexeme, Value: value, Segments: segments, Line: lexer.startLine, Column: lexer.startColumn})
}

func (lexer *Lexer) add(token Token) error {
	if len(lexer.tokens) >= lexer.config.MaxTokens {
		return lexer.failure(fmt.Sprintf("token budget exhausted at %d tokens", lexer.config.MaxTokens), ErrTokenBudget)
	}
	lexer.tokens = append(lexer.tokens, token)
	return nil
}

func (lexer *Lexer) advance() rune {
	char := lexer.source[lexer.current]
	lexer.current++
	if char == '\n' {
		lexer.line++
		lexer.column = 1
	} else {
		lexer.column++
	}
	return char
}

func (lexer *Lexer) match(expected rune) bool {
	if lexer.atEnd() || lexer.source[lexer.current] != expected {
		return false
	}
	lexer.advance()
	return true
}

func (lexer *Lexer) peek() rune {
	if lexer.atEnd() {
		return 0
	}
	return lexer.source[lexer.current]
}

func (lexer *Lexer) peekNext() rune {
	if lexer.current+1 >= len(lexer.source) {
		return 0
	}
	return lexer.source[lexer.current+1]
}

func (lexer *Lexer) atEnd() bool   { return lexer.current >= len(lexer.source) }
func (lexer *Lexer) slice() string { return string(lexer.source[lexer.start:lexer.current]) }

func (lexer *Lexer) failure(message string, cause error) error {
	return &Error{Line: lexer.startLine, Column: lexer.startColumn, Message: message, Cause: cause}
}

File: koschei/lexer.py
"""Koschei (.ks) lexical scanner."""

from __future__ import annotations

from dataclasses import dataclass
from enum import Enum, auto
from typing import Any


class TokenType(Enum):
    # Keywords
    PURE = auto()
    STATEFUL = auto()
    FN = auto()
    LET = auto()
    MUT = auto()
    OR = auto()
    RETURN = auto()
    IF = auto()
    ELSE = auto()
    WHILE = auto()
    FOR = auto()
    IN = auto()
    BREAK = auto()
    CONTINUE = auto()
    STRUCT = auto()
    ENUM = auto()
    MATCH = auto()
    IMPORT = auto()
    TRUE = auto()
    FALSE = auto()

    # Names and values
    TYPE = auto()
    IDENTIFIER = auto()
    STRING = auto()
    STRING_INTERP = auto()
    NUMBER = auto()
    COMMENT = auto()

    # Single-character symbols
    LEFT_PAREN = auto()
    RIGHT_PAREN = auto()
    LEFT_BRACE = auto()
    RIGHT_BRACE = auto()
    COMMA = auto()
    COLON = auto()
    DOT = auto()
    EQUAL = auto()
    PLUS = auto()
    MINUS = auto()
    STAR = auto()
    SLASH = auto()
    PERCENT = auto()
    SEMICOLON = auto()
    BANG = auto()
    LEFT_BRACKET = auto()
    RIGHT_BRACKET = auto()

    # Multi-character symbols
    ARROW = auto()
    FAT_ARROW = auto()
    EQUAL_EQUAL = auto()
    BANG_EQUAL = auto()
    LESS = auto()
    LESS_EQUAL = auto()
    GREATER = auto()
    GREATER_EQUAL = auto()
    AMP_AMP = auto()
    PIPE_PIPE = auto()

    EOF = auto()


@dataclass(frozen=True, slots=True)
class Token:
    type: TokenType
    value: Any
    line: int
    column: int

    def __repr__(self) -> str:
        return (
            f"Token({self.type.name:<13}, {self.value!r}, "
            f"line={self.line}, column={self.column})"
        )


class LexerError(SyntaxError):
    """Raised when Koschei source cannot be tokenized."""


class Lexer:
    KEYWORDS = {
        "pure": TokenType.PURE,
        "stateful": TokenType.STATEFUL,
        "fn": TokenType.FN,
        "let": TokenType.LET,
        "mut": TokenType.MUT,
        "or": TokenType.OR,
        "return": TokenType.RETURN,
        "if": TokenType.IF,
        "else": TokenType.ELSE,
        "while": TokenType.WHILE,
        "for": TokenType.FOR,
        "in": TokenType.IN,
        "break": TokenType.BREAK,
        "continue": TokenType.CONTINUE,
        "struct": TokenType.STRUCT,
        "enum": TokenType.ENUM,
        "match": TokenType.MATCH,
        "import": TokenType.IMPORT,
        "true": TokenType.TRUE,
        "false": TokenType.FALSE,
    }

    BUILTIN_TYPES = {
        "SystemCaps",
        "NetRoot",
        "DiskRoot",
        "EnvRoot",
        "ProcessRoot",
        "NetCaps",
        "DiskCaps",
        "DiskReadCaps",
        "EnvCaps",
        "ProcessCaps",
        "List",
        "Map",
        "String",
        "Int",
        "Float",
        "Bool",
        "Void",
        "Error",
        "Result",
        "Option",
    }

    SINGLE_CHAR_TOKENS = {
        "(": TokenType.LEFT_PAREN,
        ")": TokenType.RIGHT_PAREN,
        "{": TokenType.LEFT_BRACE,
        "}": TokenType.RIGHT_BRACE,
        ",": TokenType.COMMA,
        ":": TokenType.COLON,
        ".": TokenType.DOT,
        "+": TokenType.PLUS,
        "*": TokenType.STAR,
        "%": TokenType.PERCENT,
        ";": TokenType.SEMICOLON,
        "[": TokenType.LEFT_BRACKET,
        "]": TokenType.RIGHT_BRACKET,
    }

    ESCAPES = {
        "n": "\n",
        "r": "\r",
        "t": "\t",
        '"': '"',
        "\\": "\\",
        "{": "{",
        "}": "}",
    }

    def __init__(self, source: str, *, keep_comments: bool = False) -> None:
        self.source = source
        self.keep_comments = keep_comments
        self.start = 0
        self.current = 0
        self.line = 1
        self.column = 1
        self.start_line = 1
        self.start_column = 1
        self.tokens: list[Token] = []

    def tokenize(self) -> list[Token]:
        while not self._is_at_end():
            self.start = self.current
            self.start_line = self.line
            self.start_column = self.column
            self._scan_token()
        self.tokens.append(Token(TokenType.EOF, None, self.line, self.column))
        return self.tokens

    def _scan_token(self) -> None:
        char = self._advance()
        if char in {" ", "\r", "\t", "\n"}:
            return

        if char == "/":
            if self._match("/"):
                while self._peek() not in {"\n", "\0"}:
                    self._advance()
                if self.keep_comments:
                    self._add_token(
                        TokenType.COMMENT, self.source[self.start:self.current]
                    )
                return
            self._add_token(TokenType.SLASH, "/")
            return

        if char == '"':
            self._string()
            return
        if char.isdigit():
            self._number()
            return
        if char.isalpha() or char == "_":
            self._identifier()
            return

        if char == "-":
            if self._match(">"):
                self._add_token(TokenType.ARROW, "->")
            else:
                self._add_token(TokenType.MINUS, "-")
            return
        if char == "=":
            if self._match(">"):
                self._add_token(TokenType.FAT_ARROW, "=>")
            elif self._match("="):
                self._add_token(TokenType.EQUAL_EQUAL, "==")
            else:
                self._add_token(TokenType.EQUAL, "=")
            return
        if char == "!":
            if self._match("="):
                self._add_token(TokenType.BANG_EQUAL, "!=")
            else:
                self._add_token(TokenType.BANG, "!")
            return
        if char == "&":
            if self._match("&"):
                self._add_token(TokenType.AMP_AMP, "&&")
                return
            self._error("Beklenmeyen '&'. Mantıksal ve için '&&' kullanın.")
        if char == "|":
            if self._match("|"):
                self._add_token(TokenType.PIPE_PIPE, "||")
                return
            self._error("Beklenmeyen '|'. Mantıksal veya için '||' kullanın.")
        if char == "<":
            if self._match("="):
                self._add_token(TokenType.LESS_EQUAL, "<=")
            else:
                self._add_token(TokenType.LESS, "<")
            return
        if char == ">":
            if self._match("="):
                self._add_token(TokenType.GREATER_EQUAL, ">=")
            else:
                self._add_token(TokenType.GREATER, ">")
            return

        token_type = self.SINGLE_CHAR_TOKENS.get(char)
        if token_type is not None:
            self._add_token(token_type, char)
            return
        self._error(f"Geçersiz karakter: {char!r}")

    def _identifier(self) -> None:
        while self._peek().isalnum() or self._peek() == "_":
            self._advance()
        text = self.source[self.start:self.current]
        token_type = self.KEYWORDS.get(text)
        if token_type is None:
            token_type = (
                TokenType.TYPE
                if text in self.BUILTIN_TYPES or text[:1].isupper()
                else TokenType.IDENTIFIER
            )
        self._add_token(token_type, text)

    def _number(self) -> None:
        while self._peek().isdigit():
            self._advance()
        is_float = False
        if self._peek() == "." and self._peek_next().isdigit():
            is_float = True
            self._advance()
            while self._peek().isdigit():
                self._advance()
        text = self.source[self.start:self.current]
        self._add_token(TokenType.NUMBER, float(text) if is_float else int(text))

    def _string(self) -> None:
        segments: list[tuple[str, str]] = []
        text_parts: list[str] = []

        def flush_text() -> None:
            if text_parts:
                segments.append(("text", "".join(text_parts)))
                text_parts.clear()

        while not self._is_at_end():
            char = self._advance()
            if char == '"':
                flush_text()
                if any(kind == "expr" for kind, _ in segments):
                    self._add_token(TokenType.STRING_INTERP, tuple(segments))
                else:
                    self._add_token(
                        TokenType.STRING, segments[0][1] if segments else ""
                    )
                return
            if char == "\\":
                if self._is_at_end():
                    self._error("Tamamlanmamış kaçış dizisi.")
                escaped = self._advance()
                if escaped not in self.ESCAPES:
                    self._error(f"Geçersiz kaçış dizisi: \\{escaped}")
                text_parts.append(self.ESCAPES[escaped])
                continue
            if char == "{":
                expr_source = self._interpolation_expression()
                if not expr_source:
                    self._error("Boş interpolasyon: '{}' geçersizdir.")
                flush_text()
                segments.append(("expr", expr_source))
                continue
            if char == "}":
                self._error("Metin içinde tek '}' geçersizdir; '\\}' kullanın.")
            text_parts.append(char)
        self._error("Kapatılmamış metin değeri.")

    def _interpolation_expression(self) -> str:
        depth = 1
        expression: list[str] = []
        in_string = False
        escaped = False
        while not self._is_at_end():
            char = self._advance()
            if in_string:
                expression.append(char)
                if escaped:
                    escaped = False
                elif char == "\\":
                    escaped = True
                elif char == '"':
                    in_string = False
                elif char == "\n":
                    self._error("İnterpolasyon içindeki metin satır sonuna taşamaz.")
                continue
            if char == '"':
                in_string = True
                expression.append(char)
                continue
            if char == "{":
                depth += 1
                expression.append(char)
                continue
            if char == "}":
                depth -= 1
                if depth == 0:
                    return "".join(expression).strip()
                expression.append(char)
                continue
            if char == "\n":
                self._error("İnterpolasyon '}' ile aynı satırda kapatılmalıdır.")
            expression.append(char)
        self._error("İnterpolasyon '}' ile kapatılmalıdır.")
        raise AssertionError("unreachable")

    def _add_token(self, token_type: TokenType, value: Any) -> None:
        self.tokens.append(
            Token(token_type, value, self.start_line, self.start_column)
        )

    def _advance(self) -> str:
        char = self.source[self.current]
        self.current += 1
        if char == "\n":
            self.line += 1
            self.column = 1
        else:
            self.column += 1
        return char

    def _match(self, expected: str) -> bool:
        if self._is_at_end() or self.source[self.current] != expected:
            return False
        self._advance()
        return True

    def _peek(self) -> str:
        return "\0" if self._is_at_end() else self.source[self.current]

    def _peek_next(self) -> str:
        if self.current + 1 >= len(self.source):
            return "\0"
        return self.source[self.current + 1]

    def _is_at_end(self) -> bool:
        return self.current >= len(self.source)

    def _error(self, message: str) -> None:
        raise LexerError(
            f"[satır {self.start_line}, sütun {self.start_column}] {message}"
        )


def tokenize(source: str, *, keep_comments: bool = False) -> list[Token]:
    return Lexer(source, keep_comments=keep_comments).tokenize()
